src/bounding_hierarchy.rs

```diff
@@ -245,6 +247,83 @@ pub trait BoundingHierarchy<T: BHValue, const D: usize> {
         shapes: &'a [Shape],
     ) -> Vec<&'a Shape>;
 
+    /// Traverses the [`BoundingHierarchy`].
+    /// Returns the shape closest to the query point.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bvh::aabb::{Aabb, Bounded};
+    /// use bvh::bounding_hierarchy::BoundingHierarchy;
+    /// use bvh::bvh::Bvh;
+    /// use bvh::point_query::PointDistance;
+    /// use nalgebra::{Point3, Vector3};
+    /// use bvh::ray::Ray;
+    /// # use bvh::bounding_hierarchy::BHShape;
+    /// # pub struct UnitBox {
+    /// #     pub id: i32,
+    /// #     pub pos: Point3<f32>,
+    /// #     node_index: usize,
+    /// # }
+    /// #
+    /// # impl UnitBox {
+    /// #     pub fn new(id: i32, pos: Point3<f32>) -> UnitBox {
+    /// #         UnitBox {
+    /// #             id: id,
+    /// #             pos: pos,
+    /// #             node_index: 0,
+    /// #         }
+    /// #     }
+    /// # }
+    /// #
+    /// # impl Bounded<f32,3> for UnitBox {
+    /// #     fn aabb(&self) -> Aabb<f32,3> {
+    /// #         let min = self.pos + Vector3::new(-0.5, -0.5, -0.5);
+    /// #         let max = self.pos + Vector3::new(0.5, 0.5, 0.5);
+    /// #         Aabb::with_bounds(min, max)
+    /// #     }
+    /// # }
+    /// #
+    /// # impl BHShape<f32,3> for UnitBox {
+    /// #     fn set_bh_node_index(&mut self, index: usize) {
+    /// #         self.node_index = index;
+    /// #     }
+    /// #
+    /// #     fn bh_node_index(&self) -> usize {
+    /// #         self.node_index
+    /// #     }
+    /// # }
+    /// #
+    /// # impl PointDistance<f32,3> for UnitBox {
+    /// #     fn distance_squared(&self, query_point: Point3<f32>) -> f32 {
+    /// #         self.aabb().get_min_distance_2(query_point)
+    /// #    }
+    /// }
+    /// # fn create_bvh() -> (Bvh<f32,3>, Vec<UnitBox>) {
+    /// #     let mut shapes = Vec::new();
+    /// #     for i in 0..1000 {
+    /// #         let position = Point3::new(i as f32, i as f32, i as f32);
+    /// #         shapes.push(UnitBox::new(i, position));
+    /// #     }
+    /// #     let bvh = Bvh::build(&mut shapes);
+    /// #     (bvh, shapes)
+    /// # }
+    ///
+    /// let (bvh, shapes) = create_bvh();
+    ///
+    /// let query = Point3::new(5.0, 0.0, 2.0);
+    /// let nearest_shape = bvh.nearest_from(query, &shapes);
+    /// ```
+    ///
+    /// [`BoundingHierarchy`]: trait.BoundingHierarchy.html
+    /// [`Aabb`]: ../aabb/struct.Aabb.html
+    ///
+    fn nearest_from<'a, Shape: BHShape<T, D> + PointDistance<T, D>>(
+        &'a self,
+        query: Point<T, D>,
+        shapes: &'a [Shape],
+    ) -> Option<(&'a Shape, T)>;
```

Comment:
IMO it's worth calling this in the fuzzer (`fuzz/`) for a normal BVH and a flat BVH, even if the output is disregarded, to make sure it doesn't panic. A simple and easy check would be to make sure it returns `None` if and only if the BVH is empty at that moment.

If you are unsure about this, I'm happy to implement it myself in a followup PR.